prereise/gather/demanddata/transportation_electrification/data_process.py

```diff
@@ -0,0 +1,132 @@
+import pandas as pd
+
+#Download CSV of NHTS 2017 from https://nhts.ornl.gov/
+
+def data_filtering(census_division, filepath: str = "trippub.csv"):
+    """Filter raw NHTS data to be used in mileage.py
+    
+    :param str filepath: filepath for the "trippub.csv" file in downloaded NHTS data
+    :return: (*pd.DataFrame*) -- filtered and sorted trip data
+    """
+
+    nhts_raw = pd.read_csv(filepath)
+    
+    #filter to be only vehicle trips (TRPTRANS values 1-6)
+    nhts = nhts_raw[nhts_raw.TRPTRANS.isin(range(1,7))]
+    
+    #filter out repeated trips (VMT = -1)
+    nhts = nhts[nhts.TRPMILES != -1]
+    
+    #get correct census division
+    nhts = nhts[nhts.CENSUS_D == census_division]
+
+    #dataframe for use built here
+    #columns taken directly from original csv:
+    sorted_data = pd.DataFrame({"Household": nhts["HOUSEID"], 
+                   "Vehicle ID": nhts["VEHID"], 
+                   "Person ID": nhts["PERSONID"],
+                   "Scaling Factor Applied": nhts["WTTRDFIN"],
+                   "Trip Number": nhts["TDTRPNUM"],
+                   "Date": nhts["TDAYDATE"], 
+                   "Day of Week": nhts["TRAVDAY"],              
+                   "If Weekend": nhts["TDWKND"],                
+                   "Trip start time (HHMM)": nhts["STRTTIME"],
+                   "Trip end time (HHMM)": nhts["ENDTIME"],
+                   "Travel Minutes": nhts["TRVLCMIN"],
+                   "Dwell time": nhts["DWELTIME"],
+                   "Miles traveled": nhts["VMT_MILE"],          
+                   "Vehicle miles travelled": nhts["TRPMILES"], 
+                   "why from": nhts["WHYFROM"],
+                   "why to": nhts["WHYTO"],
+                   "Vehicle type": nhts["TRPTRANS"],            
+                   "Household vehicle count": nhts["HHVEHCNT"],
+                   "Household size": nhts["HHSIZE"]
+                   })
+    
+    #columns that require calculations: 21-28
+    sorted_data["Start time (hour decimal)"] = [row//100 + row % 100 / 60 for row in sorted_data["Trip start time (HHMM)"]]
+    sorted_data["End time (hour decimal)"] = [row//100 + row % 100 / 60 for row in sorted_data["Trip end time (HHMM)"]]
+    sorted_data['Dwell time (hour decimal)'] = ''
+    sorted_data["Travel time (hour decimal)"]= sorted_data["End time (hour decimal)"] \
+        - sorted_data["Start time (hour decimal)"]
+    sorted_data["Vehicle speed (mi/hour)"] = sorted_data["Vehicle miles travelled"] \
+        / sorted_data["Travel time (hour decimal)"]
+    sorted_data['sample vehicle number'] = ''
+    sorted_data['total vehicle trips'] = ''
+    sorted_data['total vehicle miles travelled'] = ''
+
+
+    i = 0
+    l = len(sorted_data)
+    curr_household = sorted_data.iloc[0, sorted_data.columns.get_loc('Household')]
+    curr_vehicle_id = sorted_data.iloc[0, sorted_data.columns.get_loc('Vehicle ID')]
+    sample_veh_num = 1
+    total_trips = 0
+    total_miles = 0
+
+    while i < l:
```

Comment:
As @limingzhou2004 points out, we can also combine the calculation of group totals with the assignment back to individual rows:
```python
total_trips = grouping.size()
df["total_trips"] = df["class"].map(total_trips)
total_miles = grouping["miles"].sum()
df["total_miles"] = df["class"].map(total_miles)
```
can become:
```python
df["total_trips"] = grouping["miles"].transform(len)
df["total_miles"] = grouping["miles"].transform(sum)
```
This will avoid us having to define a series of `keys` values.